server/src/main/java/org/apache/druid/server/coordinator/helper/DruidCoordinatorBalancer.java

```diff
@@ -107,25 +107,25 @@ private void balanceTier(
     }
 
     /*
-      Take as much segments from maintenance servers as priority allows and find the best location for them on
+      Take as many segments from decommissioned servers as priority allows and find the best location for them on
       available servers. After that, balance segments within available servers pool.
      */
     Map<Boolean, List<ServerHolder>> partitions =
-        servers.stream().collect(Collectors.partitioningBy(ServerHolder::isInMaintenance));
-    final List<ServerHolder> maintenanceServers = partitions.get(true);
+        servers.stream().collect(Collectors.partitioningBy(ServerHolder::isDecommissioned));
+    final List<ServerHolder> decommssionedServers = partitions.get(true);
     final List<ServerHolder> availableServers = partitions.get(false);
     log.info(
-        "Found %d servers in maintenance, %d available servers servers",
-        maintenanceServers.size(),
+        "Found %d decomissioned servers, %d available servers servers",
+        decommssionedServers.size(),
         availableServers.size()
     );
 
-    if (maintenanceServers.isEmpty()) {
+    if (decommssionedServers.isEmpty()) {
       if (availableServers.size() <= 1) {
         log.info("[%s]: %d available servers servers found.  Cannot balance.", tier, availableServers.size());
       }
     } else if (availableServers.isEmpty()) {
-      log.info("[%s]: no available servers servers found during maintenance.  Cannot balance.", tier);
+      log.info("[%s]: no available servers servers found during decommissioning.  Cannot balance.", tier);
```

Comment:
Suggested to change the message to something like "no active servers found, segments can't be moved off %d decommissioning servers"

 - Maybe it should be `warn()`, not `info()`?
 - Seems that this logging statement and the statement at line 125 duplicate the statement at line 117. Maybe this part of code should be restructured.

Reply:
"no active servers found, segments can't be moved off %d decommissioning servers" leads me to the question: should we really give up decommission in this case? Maybe if segments are available on historical nodes from other tiers (or from the same tier, on other decommissioning servers or on full active servers -- see below), or actually always, we should continue decommission by just dropping segments from the decommissioning servers and not moving them anywhere?

Also this condition doesn't cover the situation when there are active historical nodes in the tier, but they are either full, or their load queuess have `maxSegmentsInNodeLoadingQueue` elements (loading segments). (Note that in the latter case, balancing step won't be skipped entirely in the `if (!currentlyMovingSegments.get(tier).isEmpty()) {` branch above in the code, because historicals' segment loading queues may be filled up by `LoadRule`, because `LoadRule` uses an _independent_ mechanism to implement the "making a balancing burst and then letting all segments to be loaded before the next balacing burst" pattern, `ReplicationThrottler`).

Reply:
I elaborated the idea of the problems related to the independence of `
DruidCoordinatorBalancer` and `LoadRule` in #7159.

Reply:
> "no active servers found, segments can't be moved off %d decommissioning servers" leads me to the question: should we really give up decommission in this case? Maybe if segments are available on historical nodes from other tiers (or from the same tier, on other decommissioning servers or on full active servers -- see below), or actually always, we should continue decommission by just dropping segments from the decommissioning servers and not moving them anywhere?

@egor-ryashin, what's your take on this?

Reply:
@leventov the initial idea was to keep the replication at the same level.

Reply:
Consolidated the 'cannot balance' checks here and changed the message here to just log.warn that there is an insufficient number of active servers to do anything and added a `return`

Reply:
> the initial idea was to keep the replication at the same level.

This should be reflected in docs, as well as other corner cases related to decommissioning nodes.

Reply:
I'll try to add docs that decommissioning can become stalled if there are no active servers available to move segments to.